File: solutions/0200-number-of-islands/solution.cpp
class Solution {
public:

    void color(vector<vector<char>>& grid,int i,int j){
        if(i>=grid.size() or i<0 or j<0 or j>=grid[0].size()){
            return;
        }
        if(grid[i][j]=='0' or grid[i][j]=='2'){
            return;
        }
        if(grid[i][j]=='1'){
            grid[i][j] = '2';
        }
        color(grid,i,j+1);
        color(grid,i,j-1);
        color(grid,i+1,j);
        color(grid,i-1,j);
    }

    int numIslands(vector<vector<char>>& grid) {
        int count_of_color_initiation =0;

        // find 1 and start coloring(negative 1)
        int i=0;
        int j=0;

        for(int i=0;i<grid.size();i++){
            for(int j=0;j<grid[0].size();j++){
                if(grid[i][j]=='1'){
                    color(grid,i,j);
                    count_of_color_initiation++;
                }
            }
        }

        return count_of_color_initiation;
    }
};

File: solutions/0200-number-of-islands/solution.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        count  = 0
        n = len(grid)
        m = len(grid[0])
        visited = [[0 for _ in range(m)] for _ in range(n)]

        def traverse(i,j):
            if(i>=n or j>=m or i<0 or j<0):
                return
            if(visited[i][j]==1):
                return 
            if(grid[i][j]=="0"):
                return
            visited[i][j]=1
            # go near 4
            traverse(i+1,j)
            traverse(i,j+1)
            traverse(i-1,j)
            traverse(i,j-1)

        for i in range(n):
            for j in range(m):
                if(visited[i][j]==0 and grid[i][j]=="1"):
                    traverse(i,j)
                    count+=1
        return count
        
